# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [4]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [5]:
# Create our features
X= df.copy()
X.drop('loan_status',axis=1, inplace=True)
print(X.shape)
# check the data type of X and get the columns with 'object' data type
column_names = X.dtypes[X.dtypes=='object'].index

# Encoding columns of X to binary data
X = pd.get_dummies(X, columns=column_names)

# Create our target
y = df['loan_status']


(68817, 85)


In [6]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Mar-2019  pymnt_plan_n  \
count   68817.000000  ...      68817.000000       68817.0   
mean    17604.142828  ...          0.177238           1.0   
std     21835.880400  ...          0.381873           0.0   
min         0.000000  ...          0.000000           1.0   
25%      6293.000000  ...          0.000000           1.0   
50%     12068.000000  ...          0.000000           1.0   
75%     21735.000000  ...          0.000000           1.0   
max    587191.000000  ...          1.000000           1.0   

       initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.123879               0.876121               0.383161   
std                 0.329446               0.329446               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 0.000000               1.000000               0.000000   
50%                 0.000000               1.000000               0.000000   
75%                 0.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       next_pymnt_d_May-2019  application_type_Individual  \
count           68817.000000                 68817.000000   
mean                0.616839                     0.860340   
std                 0.486161                     0.346637   
min                 0.000000                     0.000000   
25%                 0.000000                     1.000000   
50%                 1.000000                     1.000000   
75%                 1.000000                     1.000000   
max                 1.000000                     1.000000   

       application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
count                68817.000000          68817.0                 68817.0  
mean                     0.139660              1.0                     1.0  
std                      0.346637              0.0                     0.0  
min                      0.000000              1.0                     1.0  
25%                      0.000000              1.0                     1.0  
50%                      0.000000              1.0                     1.0  
75%                      0.000000              1.0                     1.0  
max                      1.000000              1.0                     1.0  

[8 rows x 95 columns]

In [7]:
# Check the balance of our target values
Counter(y)

Counter({'low_risk': 68470, 'high_risk': 347})

In [8]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)
X_train.shape

(51612, 95)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [9]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

In [10]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [11]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [12]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
balanced_rf_model = BalancedRandomForestClassifier(n_estimators=500, random_state=1)

# Fit the model
balanced_rf_model = balanced_rf_model.fit(X_train_scaled, y_train)

# Making predictions using testing data
y_pred_brf = balanced_rf_model.predict(X_test_scaled)


In [13]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test,y_pred_brf)

0.8025107670490026

In [14]:
# Display the confusion matrix
confusion_matrix(y_test,y_pred_brf)

array([[   59,    28],
       [ 1252, 15866]], dtype=int64)

In [15]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_brf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.05      0.68      0.93      0.08      0.79      0.61        87
   low_risk       1.00      0.93      0.68      0.96      0.79      0.64     17118

avg / total       0.99      0.93      0.68      0.96      0.79      0.64     17205



In [16]:
# List the features sorted in descending order by feature importance
# Get the feature importance array
importances = balanced_rf_model.feature_importances_

# List the features in descending order
importances_sorted = sorted(zip(balanced_rf_model.feature_importances_, X.columns),reverse=True)
importances_sorted

[(0.06502785730329358, 'total_rec_prncp'),
 (0.06480222129516051, 'total_rec_int'),
 (0.06256804840515041, 'total_pymnt_inv'),
 (0.061530359988374285, 'total_pymnt'),
 (0.05397128419817156, 'last_pymnt_amnt'),
 (0.02687378057796136, 'int_rate'),
 (0.018936354248845554, 'issue_d_Jan-2019'),
 (0.0168517491962738, 'dti'),
 (0.016785433008545916, 'installment'),
 (0.016544877702872135, 'out_prncp'),
 (0.016370276321787138, 'out_prncp_inv'),
 (0.016263300913634298, 'annual_inc'),
 (0.015885453125281113, 'mths_since_recent_inq'),
 (0.015774921185452907, 'bc_util'),
 (0.015773557261803326, 'revol_bal'),
 (0.01576109275345071, 'max_bal_bc'),
 (0.014608619017441753, 'mo_sin_old_rev_tl_op'),
 (0.014519173490957625, 'mo_sin_old_il_acct'),
 (0.014518699060077368, 'avg_cur_bal'),
 (0.013866393242974505, 'il_util'),
 (0.013779901376596689, 'issue_d_Mar-2019'),
 (0.013764862545947986, 'total_bc_limit'),
 (0.013699765733465473, 'tot_cur_bal'),
 (0.01365758321979286, 'bc_open_to_buy'),
 (0.013214343206

### Easy Ensemble Classifier

In [17]:
from imblearn.ensemble import EasyEnsembleClassifier

# Create the easy ensemble classifier instance
ee_model = EasyEnsembleClassifier(n_estimators= 500, random_state=1)

# Fit the model
ee_model = ee_model.fit(X_train_scaled, y_train)

# Making predictions using the testing data
y_pred_ee = ee_model.predict(X_test_scaled)


In [18]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test, y_pred_ee)

0.9263620468069506

In [19]:
# Display the confusion matrix
confusion_matrix(y_test,y_pred_ee)

array([[   79,     8],
       [  947, 16171]], dtype=int64)

In [20]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_ee))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.08      0.91      0.94      0.14      0.93      0.85        87
   low_risk       1.00      0.94      0.91      0.97      0.93      0.86     17118

avg / total       0.99      0.94      0.91      0.97      0.93      0.86     17205



### Final Questions

1. Which model had the best balanced accuracy score?

  Easy Ensemble Model had the best balanced accuracy score 0.9263620468069506

2. Which model had the best recall score?

  Easy Ensemble Model had the best recall score. 0.91 for high risk and 0.94 for low risk.

3. Which model had the best geometric mean score?

  Easy Ensemble Model had the best geometric mean score 0.93

4. What are the top three features?

  Top three features are: total_rec_prncp, total_rec_int, and total_pymnt_inv.